Document virtual keyboard feature integration for Android application (#1)
anbox-cloud/discourse-docs: client-side-virtual-keyboard.md

# Client Side Virtual Keyboard

> Anbox Streaming SDK enables developers to build a hybrid mobile application that can integrate rich features that Anbox Cloud provides. It comes with an [Android library](https://developer.android.com/studio/projects/android-library) which offers easy-to-use native components like AnboxWebView which helps to quickly create client side virtual keyboard feature integrated mobile client that can send the text to Android container on the fly when typing. It extends the AOSP [WebView](https://developer.android.com/reference/android/webkit/WebView) and provides better handling of the text input for the hybrid application that loads the Anbox Streaming JavaScript SDK with an embedded WebView for video streaming.

Author: That looks good to me.
I prefer to use
`...which you can use to quickly integrate a client-side virtual keyboard feature into your mobile client..`
over
`...which you can use to quickly integrate a client-side virtual keyboard feature into your mobile application..` to avoid double `client` terms in one sentence.
How do you feel about that?

> Anbox Streaming SDK enables developers to build a hybrid mobile application that can integrate rich features that Anbox Cloud provides. It comes with an [Android library](https://developer.android.com/studio/projects/android-library) which offers easy-to-use native components like AnboxWebView which helps to quickly create client side virtual keyboard feature integrated mobile client that can send the text to Android container on the fly when typing. It extends the AOSP [WebView](https://developer.android.com/reference/android/webkit/WebView) and provides better handling of the text input for the hybrid application that loads the Anbox Streaming JavaScript SDK with an embedded WebView for video streaming.

Reviewer: Yes, that sounds better. Thanks!

## Import the AAR library

Check out the [Anbox Streaming SDK](https://github.com/anbox-cloud/anbox-streaming-sdk) from Github

```
$ git clone https://github.com/anbox-cloud/anbox-streaming-sdk.git

> ```

Author: +1. Will fix

> An AAR file `com.canonical.anbox.streaming_sdk.aar` can be found in the `android/libs` folder. Please refer to the official [documentation](https://developer.android.com/studio/projects/android-library)  on how to import an Android library into an Android application project. Also under the `examples/android` folder, the `enhanced_webview_streaming` Android application can be referred as an example for this feature integration.

Author: +1, Will fix.

## Integrate components

> After importing the AAR file, to use the AnboxWebView in an Android project

Author: +1, Will fix

> 1. Adjust the layout of the activity xml file that you want to use the AnboxWebView component, E.g:

Author: +1, `for` sounds natural to me.

   ```
    <com.canonical.anbox.streaming_sdk.AnboxWebView
        android:id="@+id/webview"
        android:layout_width="match_parent"
        android:layout_height="match_parent" />
   ```

> 2. Utilize the [addJavascriptInterface](https://developer.android.com/reference/android/webkit/WebView#addJavascriptInterface) method, inject a Java object `AppIntercae` to the AnboxWebView acts as a bridge which provides access from the JavaScript to Android Java Layer.

Author: Fixed

   ```
   import com.canonical.anbox.streaming_sdk.AnboxWebView;
   ...

   public class MainActivity extends AppCompatActivity {
       protected void onCreate(Bundle savedInstanceState) {
           ...
           mWebview = (AnboxWebView) findViewById(R.id.webview);
           mWebView.addJavascriptInterface(new AppInterface(this), "AppInterface");
           ...
       }
   }
   ```

> 3. For the application running in the Android container, whenever a text editor gains or loses focus during streaming, an IME specific message will be sent from Android container to the client through Javascript SDK to request one certain IME operation. This allows you to perform one certain action when the message is received and ensures the behavior of the virtual keyboard is synced on both ends. The supported IME messages are as follows:

Author: Fixed

               messageReceived: (type, data) => {
                   switch(type) {
                   case "show-ime" :
                       AppInterface.openVirtualKeyboard();
                   break;
                   case "hide-ime" :
                       AppInterface.hideVirtualKeyboard();
                   break;
                   default:
                       console.log("Unhandled message:" + type)
                   break;
                   }
               }
           }
       });
   ```

4. Setup the listener to the activity where you want to capture the message. The following methods need to be implemented for AppInterface.ActionListener interfaces so that the application will be notified when the request is coming from JavaScript SDK.

   ```
   public class MainActivity extends AppCompatActivity implements AppInterface.ActionListener {
   ...

       /**
        * Called as the request to open virtual keyboard
        */
       @Override
       public void onOpenVirtualKeyboard() {
       }

       /**
        * Called as the request to hide virtual keyboard
        */
       @Override
       public void onHideVirtualKeyboard() {
       }

   ```

> 5. Set the listener to the activity that you want to capture the text input events from the virtual keyboard or monitor its visibility changes during streaming.

Reviewer: ```suggestion
5. Set the listener to the activity for which you want to capture the text input events from the virtual keyboard or monitor its visibility changes during streaming.
```

   ```
   import com.canonical.anbox.streaming_sdk.AnboxWebView;
   ...

   public class MainActivity extends AppCompatActivity implements AnboxWebView.VirtualKeyboardListener, AppInterface.ActionListener {
   ...
       protected void onCreate(Bundle savedInstanceState) {
           ...
           mWebview = (AnboxWebView) findViewById(R.id.webview);
           mWebview.setVirtualKeyboardListener(this);
           ...
       }
   ```

>    When people start typing, one of the following methods from AnboxWebView.VirtualKeyboardListener interfaces will be triggered.

Reviewer: ```suggestion
   When people start typing, one of the following methods from the AnboxWebView.VirtualKeyboardListener interfaces will be triggered:
```

        */
       @Override
       public void onVirtualKeyboardTextDeleted(int counts) {
       }

       /**
        * Called as the region of composing text is changed.
        *
        * @param start the position in the text at which the composing region begins.
        * @param end the position in the text at which the composing region ends.
        */
       @Override
       public void onVirtualKeyboardComposingTextRegionChanged(int start, int end) {
       }

       /**
        * Called as the state of the virtual keyboard is changed.
        *
        * @param visible the visibility of the virtual keyboard, true or false.
        * @param heightRatio the ratio of virtual keyboard's height to screen when virtual keyboard is visible.
        *        This can be used to notify the IME running in the Android container
        *        to adjust the display height to honor the virtual keyboard display
        *        ratio on the client side.
        *
        */
       @Override
       public void onVirtualKeyboardStateChanged(boolean visible, double displayRatio) {
       }

   ```

> 6. In HTML, Defines the JavaScript functions that can be invoked from Android Java layer(AnboxWebView). Those functions are just wrappers of the APIs that are exposed from the JavaScript SDK. This builds up a communication tunnel that works the other way around.

Reviewer: ```suggestion
6. In your HTML, define the JavaScript functions that can be invoked from the Android Java layer(AnboxWebView). Those functions are just wrappers of the APIs that are exposed from the JavaScript SDK. This builds up a communication tunnel that works the other way around.
```

Author: +1, Will fix

   ```
   <script>
     function sendIMECommittedText(text) {
       window.stream.sendIMECommittedText(text);
     }
     function sendIMEComposingText(text) {
       window.stream.sendIMEComposingText(text);
     }
     function sendIMETextDeletion(counts) {
       window.stream.sendIMETextDeletion(counts);
     }
     function sendIMEAction(name, params) {
       window.stream.sendIMEAction(name, params);
     }
     function sendIMEComposingRegion(start, end) {
       window.stream.sendIMEComposingRegion(start, end);
     }
   </script>

   ```

> 7. Once the virtual keyboard pops up on the client side, the `onVirtualKeyboardStateChanged` callback function will be triggered, to keep display proportions correct for IME displaying on both client and server ends, the action with the name `show` that carries display ratio needs to be sent out to the server side.

Reviewer: ```suggestion
7. Once the virtual keyboard pops up on the client side, the `onVirtualKeyboardStateChanged` callback function is triggered. To keep the display proportions correct for IME displaying on both the client and the server, the `show` action, which carries the display ratio, must be sent out to the server side.
```

Author: +1, Will fix

   ```
       @Override
       public void onVirtualKeyboardStateChanged(boolean visible, double displayRatio) {
           if (visible == true) {
               String action = "show";
               String params = "height-ratio=" + displayRatio;
               mWebView.loadUrl(String.format("javascript:sendIMEAction(\"%s\", \"%s\")", action, params));
           }
       }

   ```

> 8. While typing on the virtual keyboard, a text input event will be triggered when one of the following scenarios occurs:
>    - A bunch of characters is committed to a text editor
>    - A text is currently in composing
>    - A text is being deleted

Reviewer: ```suggestion
8. When typing on the virtual keyboard, a text input event is triggered when one of the following scenarios occurs:
   - A series of characters is committed to a text editor
   - A text is currently being composed
   - A text is being deleted
```

Author: +1, Will fix

>     In the above cases, the changed text must be sent to the server side through the JS SDK. This can be done by calling JS functions that are defined in HTML through AnboxWebView. E.g. for the case of text commitment.

Reviewer: ```suggestion
    In the above cases, the changed text must be sent to the server side through the JavaScript SDK. This can be done by calling JavaScript functions that are defined in HTML through AnboxWebView. For example, for committing text:
```

Author: +1, Will fix

   ```
       @Override
       public void onVirtualKeyboardTextCommitted(String text) {
           mWebView.loadUrl(String.format("javascript:sendIMECommittedText(\"%s\")", text));
       }

   ```

> The above workflow provides an general overview of client side virtual keyboard feature integration into an Android application. For the complete implementation details, please refer to `enhanced_webview_streaming` example from `example/android` folder.

Reviewer: ```suggestion
The above workflow provides a general overview of how to integrate the client-side virtual keyboard feature integration into an Android application. For the complete implementation details, refer to the `enhanced_webview_streaming` example in the `example/android` folder.
```

Author: +1, Will fix
